### PR TITLE
Implements Commands Key mapping

### PR DESCRIPTION
**Pull request by ppot**
-> keymaps files
-> config structure change (file deconstruction)
-> plugins extends keymaps
-> update hterm to use `commands-registry`

### CODE REVIEW

File: app/config.js [hunk 1]
```diff
@@ -1,82 +1,31 @@
...
-  }
-});
-
-const path = _paths.confPath;
-const pathLegacy = _paths.pathLegacy;
+const win = require('./config/windows');
 
 const watchers = [];
-
+const scanInterval = 2000;
```
Comment (albinekb) [inline app/config.js]: I would:
```js
const getWatchConfig = () => {
  if (process.platform === 'win32') {
    // watch for changes on config every 2s on windows
    // https://github.com/zeit/hyper/pull/1772
    return { interval: 2000 }
  }

  return {}
}

const watchConfig = getWatchConfig()

...

gaze(confPath, watchConfig, function (err) {
```

easier to read?

**Comment by albinekb** [comment]
One thing that would be 💯^10 is to be consistent with variable names, I'm thinking of the words
`cfg`, `conf` and `config` that are mixed, it's easier to read and understand the code if you use the same name for these, chose one and stick to it 👌

File: app/config.js (hunk 1 above)
Comment (ppot) [inline app/config.js]: Yeah  agree on that!

File: app/config.js [hunk 2]
```diff
@@ -1,82 +1,31 @@
...
       throw err;
     }
     this.on('changed', () => {
-      try {
-        if (exec(readFileSync(path, 'utf8'))) {
-          notify('Hyper configuration reloaded!');
-          watchers.forEach(fn => fn());
-        }
-      } catch (err) {
-        dialog.showMessageBox({
```
Comment (albinekb) [inline app/config.js]: This PR changes from displaying a dialog to just a notification, correct?

File: app/config.js [hunk 3]
```diff
@@ -135,14 +54,20 @@ exports.getPlugins = function () {
...
-  recordState(win) {
-    winCfg.set('windowPosition', win.getPosition());
-    winCfg.set('windowSize', win.getSize());
+exports.getKeymaps = function () {
+  return cfg.keymaps;
+};
+
+exports.extendKeymaps = function (keymaps) {
+  if (keymaps) {
+    cfg.keymaps = keymaps;
```
Comment (albinekb) [inline app/config.js]: Hmm.. I don't like mutating stuff like this. But I think we do this in other places? Maybe that's a job for another PR.

File: app/config/import.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+const {writeFileSync, readFileSync} = require('fs');
+const {defaultConfig, confPath} = require('./paths');
+const _init = require('./init');
+const _keys = require('./keymaps');
```
Comment (albinekb) [inline app/config/import.js]: nit: rename this to `_keymaps`, easier to understand the code

File: app/config.js (hunk 3 above)
Comment (ppot) [inline app/config.js]: yeah we do this alot...

File: app/config/init.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+const _extract = function (script) {
+  const module = {};
+  script.runInNewContext({module});
+  if (!module.exports) {
+    throw new Error('Error reading configuration: `module.exports` not set');
+  }
+  return module.exports;
+};
+
+const _syntaxValidation = function (cfg) {
```
Comment (albinekb) [inline app/config/init.js]:
```js
const _syntaxValidation = function (cfg) {
    try {
      return new vm.Script(cfg, { filename: '.hyper.js', displayErrors: true, lineOffset: 0 })
    } catch (error) {
      notify(`Error loading config: ${error.name}, see DevTools for more info`)
      console.error('Error loading config:', error)
      return
    }
}
```
tested:
```js
_syntaxValidation(`
const c = 'a'
const b = 'd'
const c = '2'
`)
```

notify: *Error loading config: SyntaxError, see DevTools for more info*
console:
```js
 Error loading config:
 .hyper.js:4
const c = '2'
          ^^^
SyntaxError: Identifier 'c' has already been declared

File: app/config.js (hunk 3 above)
Comment (albinekb) [inline app/config.js]: I'm OK with this for now, but we should clean it all up

File: app/config/init.js (hunk 5 above)
Comment (ppot) [inline app/config/init.js]: Are you commenting as an error or a test?

File: app/config/paths.js [hunk 6]
```diff
@@ -15,10 +14,14 @@ let confDir = homeDir;
 const devDir = resolve(__dirname, '../..');
 const devConfig = resolve(devDir, conf);
 const defaultConfig = resolve(__dirname, defaultConf);
-const pathLegacy = resolve(homeDir, legacyConf);
 
 const icon = resolve(__dirname, '../static/icon.png');
 
+const keymapPath = resolve(__dirname, '../keymaps');
+const darwinKeys = resolve(keymapPath, 'darwin.json');
```
Comment (albinekb) [inline app/config/paths.js]: Should't these be `path.join`?
http://tips.tutorialhorizon.com/2017/05/01/path-join-vs-path-resolve-in-node-js/

don't know how resolve vs join is performance/resource-wise, 
it's probably so tiny that it's non-existent 😄  

🤔

File: app/config/keymaps.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+const _setCommandsForKeys = function (commands) {
+  for (const command in commands) {
+    if (command) {
+      keys[commands[command]] = command;
+    }
+  }
+};
+
+const _import = function (customsKeys) {
+  const path = () => {
```
Comment (albinekb) [inline app/config/keymaps.js]: Why not move this to `app/config/paths.js` and just expose it as `defaultPlatformKeyPath`?

Is there a need to get other platforms' keys?

File: app/keymaps/darwin.json [hunk 8]
```diff
@@ -0,0 +1,34 @@
+{
```
Comment (albinekb) [inline app/keymaps/darwin.json]: run these trough http://jsonprettyprint.com or similar 😄

File: app/menus/menu.js [hunk 9]
```diff
@@ -8,18 +10,19 @@ const helpMenu = require('./menus/help');
...
 module.exports = (createWindow, updatePlugins) => {
+  const commands = getKeymaps().commands;
   const menu = [].concat(
-    shellMenu(createWindow),
-    editMenu(),
-    viewMenu(),
-    pluginsMenu(updatePlugins),
-    windowMenu(),
-    helpMenu()
+    shellMenu(commands, createWindow),
```
Comment (albinekb) [inline app/menus/menu.js]: ES6 style:
```js
const menu = [
  ...shellMenu(commands, createWindow),
  ...editMenu(commands),
  ...viewMenu(commands),
  ...pluginsMenu(commands, updatePlugins),
  ...windowMenu(commands),
  ...helpMenu(commands)
]

Comment (albinekb) [inline app/menus/menu.js]: then remove this further down:
```
if (process.platform === 'darwin') {
  menu.unshift()...
```

and:
```js
const menu = [
  ...(process.platform === 'darwin' ? darwinMenu(commands) : []),
  ...shellMenu(commands, createWindow),
  ...editMenu(commands),
  ...viewMenu(commands),
  ...pluginsMenu(commands, updatePlugins),
  ...windowMenu(commands),
  ...helpMenu(commands)
]
```
💅

File: app/menus/menus/shell.js [hunk 10]
```diff
@@ -29,7 +27,7 @@ module.exports = function (createWindow) {
       },
       {
         label: 'Split Vertically',
-        accelerator: accelerators.splitVertically,
+        accelerator: commands['pane:vertical'],
```
Comment (albinekb) [inline app/menus/menus/shell.js]: maybe `pane:splitVertical` to make it clearer, same with horizontal

File: app/menus/menus/window.js [hunk 11]
```diff
@@ -1,15 +1,14 @@
...
     role: 'window',
     submenu: [
       {
         role: 'minimize',
-        accelerator: accelerators.minimize
+        accelerator: commands['window:minimize']
       },
       {
-        role: 'zoom'
+        role: 'zoom',
```
Comment (albinekb) [inline app/menus/menus/window.js]: What does this actually do? 😂 never used it

File: app/menus/menus/window.js [hunk 12]
```diff
@@ -71,7 +70,7 @@ module.exports = function () {
       },
       {
         role: 'togglefullscreen',
-        accelerators: accelerators.enterFullScreen
+        accelerators: commands['window:full']
```
Comment (albinekb) [inline app/menus/menus/window.js]: `window:toggleFullScreen`, easier to understand

File: app/config/init.js (hunk 5 above)
Comment (albinekb) [inline app/config/init.js]: haha the comment was to update _syntaxValidation to the code i wrote in my comment,
the blocks under is to show what it does

File: app/config/paths.js (hunk 6 above)
Comment (MartyGentillon) [inline app/config/paths.js]: given that they contain `..` resolve is probably better.

File: app/menus/menus/window.js (hunk 11 above)
Comment (MartyGentillon) [inline app/menus/menus/window.js]: Apparently nothing on Windows?

File: app/config/paths.js (hunk 6 above)
Comment (albinekb) [inline app/config/paths.js]: they do not @MartyGentillon `resolve(keymapPath, 'darwin.json')` `keymapPath` is already expanded

File: app/menus/menus/window.js (hunk 11 above)
Comment (albinekb) [inline app/menus/menus/window.js]: Aha, makes the window fill the screen (on macOS) 😄

File: app/config/paths.js (hunk 6 above)
Comment (MartyGentillon) [inline app/config/paths.js]: True enough for this particular path, but that would be the reasoning nonetheless.

File: app/config/init.js (hunk 5 above)
Comment (ppot) [inline app/config/init.js]: Done 👍

File: app/keymaps/darwin.json (hunk 8 above)
Comment (ppot) [inline app/keymaps/darwin.json]: Done 👍

File: app/config.js (hunk 2 above)
Comment (ppot) [inline app/config.js]: I won't change it since using a dialog prevent hyper from opening.

File: app/menus/menus/window.js (hunk 11 above)
Comment (MartyGentillon) [inline app/menus/menus/window.js]: Given that it does nothing on Windows (and KDE), should it even be in the menus on Windows (or Linux)?

Comment (MartyGentillon) [inline app/menus/menus/window.js]: Alternately, that could be a bug, It looks like that is supposed to be the same as "Maximize" in windows and linux, which don't appear anywhere in the menus on either (though they do appear in the window controls).

**Comment by ppot** [comment]
@albinekb Next review?

**Comment by TalAmuyal** [comment]
Will there be an option of completely disabling keybindings?

**Comment by ppot** [comment]
@TalAmuyal What do you mean exactly? Removing all the key from the menu?

**Comment by TalAmuyal** [comment]
I mean that as a user of `tmux` and `(neo)vim`, I have all the "on-demand" functionality in them (and the shell).
For example, some of my non-terminal keybindings:
- `<Ctrl>+T` - Open tab (aka window) -` tmux`
- `<Alt>+<Arrow>` - Move between tabs - `tmux`
- `<Ctrl>+R` - History search - `zsh`
- `<Ctrl>+W, H/J/K/L` - Switch pane - `Vim`

In short, every keybinding that the terminal provides might override an existing config.
So for me, the ultimate terminal has no keybindings, as everything is  pre-configured in a file.

**Comment by ppot** [comment]
@TalAmuyal  If you are using MacOS they are prefixed with `cmd`. You can always override them in the `.hyper.js` at the `keymaps` object.
```
  keymaps: {
    // exemple
    'window:devtools': 'ctrl+shift+o'
  }
```

**Comment by sandric** [comment]
I long waited for feature from early when I saw project, and I giving huge thx to @ppot for implementing it, so that I finally can try one more time to tackle my long lasting dream to use emacs from inside terminal. Having said that, is there any docs about how to make custom keybinding? I tried to tweak basic darwin kbs and they seems to work, but is there a way to create your own functions, not only predefined, like "editor:selectAll"?

**Comment by ppot** [comment]
Thanks @sandric  for now this could be possible but by creating plugins with commands execution function. This will be made more easily once we switch to xterm!

**Comment by Algram** [comment]
Does that mean we can use ctrl-r for reverse-i-search on linux now? If yes, how?

**Comment by ppot** [comment]
@Algram On current master yes.

**Comment by saadq** [comment]
This is awesome. Is there going to be a way to see a list of available commands, or maybe have a way to allow commands to be logged to the console when you do some action (similar to Sublime Text's `sublime.log_commands(True)`?

**Comment by ppot** [comment]
mmm Make an issue for this as a suggestion ;) something like a command invocation actions

**Comment by stereokai** [comment]
@ppot Does this fix [Command ctrl + C does not work #1121](https://github.com/zeit/hyper/issues/1121)?

**Comment by computersarecool** [comment]
So,
Given the status of this pull request, how would one enable Readlines (emacs) style keybindings in Hyperterm?

**Comment by sindrenm** [comment]
Hype! Can confirm that this is indeed working! :tada: :clap: 

Any idea when it'll get released? :grinning:

**Comment by computersarecool** [comment]
To me it is unclear from the title "Implements Commands Key mapping" what this pull request will do but it closes a lot of open issues.

@ppot will this pull request make it so that keybindings in the Hyper terminal emulator will work as they should?

i.e. will [these](http://www.keyxl.com/aaaf192/83/Linux-Bash-Shell-keyboard-shortcuts.htm) work?

**Comment by ELLIOTTCABLE** [comment]
Ditto @computersarecool, especially given that this is linked from the v1.4.0 release-notes — what, exactly, just got added? ^_^

**Comment by ynnoj** [comment]
@computersarecool @ELLIOTTCABLE https://hyper.is/#keymaps

**Comment by iamstarkov** [comment]
for @ELLIOTTCABLE and everyone else looking for available shortcuts

look 'em up here https://github.com/zeit/hyper/blob/master/app/keymaps/

**Pull request merged**